### PR TITLE
Implements Cumsum function for autograd

### PR DESCRIPTION
**Pull request by bunelr**
The ideal implementation of a backward pass for the cumsum operation would be a
cumsum operation on the reversed tensor but there is to my knowledge no reverse
operation yet in pytorch.

I needed this for one of my project so I implemented it, and I saw in issue #440 that this was a pending feature so I thought I would propose my implementation.

Let me know if there is a better place where this should be implemented, if more tests should be added or if anything is missing.

I might be able to take a look at some of the others later if no one is working on them.

### CODE REVIEW

File: torch/autograd/_functions/tensor.py [hunk 1]
```diff
@@ -591,4 +591,22 @@ def backward(self, grad_output):
...
+        self.dim = dim
+
+    def forward(self, input):
+        return torch.cumsum(input, dim=self.dim)
+
+    def backward(self, grad_output):
+        grad_sum = torch.sum(grad_output, dim=self.dim)
+
+        grad_input = torch.cumsum(-grad_output, dim=self.dim)
+        grad_input += grad_sum.expand_as(grad_input)
```
Comment (albanD) [inline torch/autograd/_functions/tensor.py]: Isn't it possible to replace this line with `grad_input -= grad_input[-1]` and remove the line summing over the `grad_output` ?

**Review by apaszke — COMMENTED** [review]
Can you please add a test for the Variable method? They're at the bottom of `test_autograd.py`.

File: torch/autograd/_functions/tensor.py (hunk 1 above)
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: I wouldn't do that. It's quite unsafe and might be correct or not, depending on the order of subtractions

Comment (bunelr) [inline torch/autograd/_functions/tensor.py]: But this would allow a useless summation over the `grad_output` tensor.

What if we:
- select the last element according to self.dim
- clone it so that it doesn't share the memory anymore
- then expand it and subtract it

This would mean removing one summation over the whole tensor, at the cost of a clone of the reduced tensor, which is smaller?

Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: Yeah that would be better!

**Comment by bunelr** [comment]
Added the method for`Variable`, as well as the associated test.

With regards to the implementation, i did a quick time measurement on the one doing the sum and the one cloning the end of the cumsum, the results are as follows:

- With an explicit sum: 0.021312472820281984
- With cloning of the end of the cumsum: 0.019439733028411864

Difference is small but it's there so i put in the fastest implementation.


For reference, measurement was done with this:

```python
import torch
from torch.autograd._functions import Cumsum

import time

nb_runs = 100
size = (1000, 1000)
inp = Variable(torch.randn(size).double(), requires_grad=True)
grad_out = torch.randn(size).double()

start = time.time()
for i in range(nb_runs):
    cum_sum_func = Cumsum(0)
    out = cum_sum_func(inp)
    out.backward(grad_out)
end = time.time()
print((end-start) / nb_runs)
```

**Comment by ngimel** [comment]
Would doing it like 
```
indices = torch.LongTensor(range(grad_output.size(dim)-1,-1,-1))
grad_input = torch.index_select(grad_output, dim, indices)
grad_input = torch.cumsum(grad_input, dim)
```
be faster?

**Comment by bunelr** [comment]
@ngimel 
This approach would indeed be preferable as its a more straightforward implementation, however it is lacking a reverse operation over the grad_input to bring back the elements in their original order now that their gradient has been computed.

I have tested it and it seems to be slower, especially when doing the cumsum along a dimension that is not the first one.


Benchmarking code:

```python
import torch
import time

from torch.autograd import Function, Variable

class CumsumClone(Function):

    def __init__(self, dim):
        super(CumsumClone, self).__init__()
        self.dim = dim

    def forward(self, input):
        return torch.cumsum(input, dim=self.dim)

    def backward(self, grad_output):
        grad_input = torch.cumsum(-grad_output, dim=self.dim)

        end_idx = grad_input.size(self.dim) - 1
        grad_sum = grad_input.select(self.dim, end_idx)
        if len(grad_output.size()) == 1:
            grad_input -= grad_sum
        else:
            grad_input -= grad_sum.unsqueeze(self.dim).expand_as(grad_input)
        grad_input += grad_output
        return grad_input

class CumsumExplicitSum(Function):

    def __init__(self, dim):
        super(CumsumExplicitSum, self).__init__()
        self.dim = dim

    def forward(self, input):
        return torch.cumsum(input, dim=self.dim)

    def backward(self, grad_output):
        grad_sum = torch.sum(grad_output, dim=self.dim)

        grad_input = torch.cumsum(-grad_output, dim=self.dim)
        grad_input += grad_sum.expand_as(grad_input)
        grad_input += grad_output
        return grad_input

class CumsumReverse(Function):

    def __init__(self, dim):
        super(CumsumReverse, self).__init__()
        self.dim = dim

    def forward(self, input):
        return torch.cumsum(input, dim=self.dim)

    def backward(self, grad_output):
        end_idx = grad_output.size(self.dim) - 1
        indices = torch.LongTensor(range(end_idx, -1, -1))
        grad_input = torch.index_select(grad_output, self.dim, indices)
        grad_input = torch.cumsum(grad_input, self.dim)
        grad_input = torch.index_select(grad_input, self.dim, indices)
        return grad_input

to_test = [CumsumClone, CumsumExplicitSum,  CumsumReverse]

nb_runs = 100
size = (100, 100, 100 )
inp = Variable(torch.randn(size).double(), requires_grad=True)
grad_out = torch.randn(size).double()

dim_to_test = range(len(size))
for dim in dim_to_test:
    for cumsum_cls in to_test:
        start = time.time()
        for i in range(nb_runs):
            cum_sum_func = cumsum_cls(dim)
            out = cum_sum_func(inp)
            grad_in = out.backward(grad_out)
        end = time.time()
        time.sleep(1)
        print("Cumsum along dim {}, Average time for {}:\t{}s per pass".format(dim,
                                                                               cumsum_cls.__name__,
                                                                               (end - start) / nb_runs))
    print()

```

Benchmarking results (desktop): 

```
$ python bench.py 
Cumsum along dim 0, Average time for CumsumClone:	0.01940817356109619s per pass
Cumsum along dim 0, Average time for CumsumExplicitSum:	0.022341022491455077s per pass
Cumsum along dim 0, Average time for CumsumReverse:	0.01978654384613037s per pass

Cumsum along dim 1, Average time for CumsumClone:	0.019124305248260497s per pass
Cumsum along dim 1, Average time for CumsumExplicitSum:	0.021526150703430176s per pass
Cumsum along dim 1, Average time for CumsumReverse:	0.02462069034576416s per pass

Cumsum along dim 2, Average time for CumsumClone:	0.012899844646453858s per pass
Cumsum along dim 2, Average time for CumsumExplicitSum:	0.015139245986938476s per pass
Cumsum along dim 2, Average time for CumsumReverse:	0.022597653865814207s per pass

```

If I rerun it several times, the trends I see are:
-> Clone is marginally faster than ExplicitSum (can be opposite on certains runs but this holds as a trend)
-> When doing the cumsum over the 0-th dimension, Reverse is in the same ballpark as the other two.
-> When doing the cumsum over another dimension, Reverse usually becomes slower than the other two


Additionally, if I do the same test over a 1D vector of size 10000, I get the following result:
```
$ python bench.py 
Cumsum along dim 0, Average time for CumsumClone:	0.0003379058837890625s per pass
Cumsum along dim 0, Average time for CumsumExplicitSum:	0.0003282761573791504s per pass
Cumsum along dim 0, Average time for CumsumReverse:	0.0019441533088684082s per pass
```

Therefore I don't think the Reverse implementation should be used.

**Comment by bunelr** [comment]
While doing the benchmarking, i discovered that were was a slight bug in my implementation, in the case of 1D tensors. I have fixed it and added a testcase for it.

File: torch/autograd/_functions/tensor.py [hunk 2]
```diff
@@ -591,4 +591,26 @@ def backward(self, grad_output):
...
+
+    def forward(self, input):
+        return torch.cumsum(input, dim=self.dim)
+
+    def backward(self, grad_output):
+        grad_input = torch.cumsum(-grad_output, dim=self.dim)
+
+        end_idx = grad_input.size(self.dim) - 1
+        grad_sum = grad_input.select(self.dim, end_idx)
+        if len(grad_output.size()) == 1:
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: It's better to use `grad_output.dim()`

File: torch/autograd/_functions/tensor.py [hunk 3]
```diff
@@ -591,4 +591,26 @@ def backward(self, grad_output):
...
+            grad_input -= grad_sum
+        else:
+            grad_input -= grad_sum.unsqueeze(self.dim).expand_as(grad_input)
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: Actually this whole if could be removed if you did that:
```python
grad_sum = grad_input.index((None,) * (self.dim - 1) + (end_idx,))
grad_input -= grad_sum.unsqueeze(self.dim).expand_as(grad_input)
```

Comment (bunelr) [inline torch/autograd/_functions/tensor.py]: I'm not familiar with the use of `None` as arguments to `index`, I could not find a reference [in the docs](http://pytorch.org/docs/tensors.html#torch.Tensor.index) and experimenting with it confused me, is it documented somewhere? (Sorry for the bother).

From experimenting, I think if you use `index`, you shouldn't `unsqueeze` at the following line.


Thanks for the help, I'll update my code.

**Comment by bunelr** [comment]
Wait, don't review yet because that's still incorrect. Sorry for the trouble.

File: torch/autograd/_functions/tensor.py [hunk 4]
```diff
@@ -591,4 +591,24 @@ def backward(self, grad_output):
...
+        self.dim = dim
+
+    def forward(self, input):
+        return torch.cumsum(input, dim=self.dim)
+
+    def backward(self, grad_output):
+        grad_input = torch.cumsum(-grad_output, dim=self.dim)
+
+        end_idx = grad_input.size(self.dim) - 1
+        index = (slice(None, None),) * self.dim + (slice(end_idx, end_idx + 1),)
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: The whole point of `index` is that you now don't need that slice. However I just realized that it might be simpler to call `grad_input.narrow(self.dim, end_idx, 1)`.

Comment (bunelr) [inline torch/autograd/_functions/tensor.py]: Indeed. I think I got confused and lost track of what I wanted to do. I'll update.

**Comment by bunelr** [comment]
This is starting to be a lot of commit for a really simple problem (Sorry about all the back and forth)

Do you want me to squash them so that the history looks cleaner?

**Comment by fmassa** [comment]
Don't worry about the squashing the commits, there is a option now in github to squash the commits prior to merging it.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks!
